src/clusterfuzz/_internal/tests/core/bot/fuzzers/centipede/centipede_engine_test.py

```diff
@@ -0,0 +1,169 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for centipede engine."""
+
+import glob
+import os
+import shutil
+import unittest
+
+from clusterfuzz._internal.bot.fuzzers import engine_common
+from clusterfuzz._internal.bot.fuzzers.centipede import engine
+from clusterfuzz._internal.system import environment
+from clusterfuzz._internal.tests.test_libs import helpers as test_helpers
+from clusterfuzz._internal.tests.test_libs import test_utils
+
+TEST_PATH = os.path.abspath(os.path.dirname(__file__))
+DATA_DIR = os.path.join(TEST_PATH, 'test_data')
+WORK_DIR = os.path.join(DATA_DIR, 'workdir')
+CORPUS_DIR = os.path.join(DATA_DIR, 'corpus')
+CRASHES_DIR = os.path.join(WORK_DIR, 'crashes')
+
+# Centipede's runtime args
+_TIMEOUT = 1200
+_SERVER_COUNT = 1
+_RSS_LIMIT = 4096
+_RLIMIT_AS = 5120
+_ADDRESS_SPACE_LIMIT = 0
+_DEFAULT_ARGUMENTS = [
+    '--exit_on_crash=1',
+    f'--timeout={_TIMEOUT}',
+    f'--fork_server={_SERVER_COUNT}',
+    f'--rss_limit_mb={_RSS_LIMIT}',
+    f'--address_space_limit_mb={_ADDRESS_SPACE_LIMIT}',
+]
+
+
+def clear_temp_dirs():
+  """Clear temp directory."""
+  for directory in [CORPUS_DIR, WORK_DIR]:
+    if os.path.exists(directory):
+      shutil.rmtree(directory)
+    os.mkdir(directory)
+
+
+def setup_testcase_and_corpus(testcase, corpus):
+  """Setup testcase and corpus."""
+  clear_temp_dirs()
+  copied_testcase_path = os.path.join(CORPUS_DIR, testcase)
+  shutil.copy(os.path.join(DATA_DIR, testcase), copied_testcase_path)
+
+  copied_corpus_path = os.path.join(CORPUS_DIR, corpus)
+  src_corpus_path = os.path.join(DATA_DIR, corpus)
+
+  if os.path.exists(src_corpus_path):
+    shutil.copytree(src_corpus_path, copied_corpus_path)
+  else:
+    os.mkdir(copied_corpus_path)
+
+  return copied_testcase_path, copied_corpus_path
+
+
+@test_utils.integration
+class IntegrationTest(unittest.TestCase):
+  """Integration tests."""
+
+  def setUp(self):
+    self.maxDiff = None  # pylint: disable=invalid-name
+    test_helpers.patch_environ(self)
+
+    os.environ['BUILD_DIR'] = DATA_DIR
+
+  def compare_arguments(self, expected, actual):
+    """Compare expected arguments."""
+    self.assertListEqual(expected, actual)
+
+  def assert_has_stats(self, results):
+    """Assert that stats exist."""
+    # Centipede does not have stats report yet.
+    # TODO(Dongge): Implement this when the feature is supported.
+
+  def test_reproduce(self):
+    """Tests reproducing a crash."""
+    testcase_path, _ = setup_testcase_and_corpus('crash', 'empty_corpus')
+    engine_impl = engine.Engine()
+    sanitized_target_path = f'{DATA_DIR}/__centipede_address/test_fuzzer'
```

Comment:
minor thing: can we rename `__centipede_address` to something even more general, like `__extra_build` ?

I'm working on build setup support, and it's a lot easier to not have to make assumptions about the sanitizer and the engine.

Reply:
Sure!

Reply:
When we want to support multiple sanitized binaries in the future, we might need to change either the binary name (e.g., `target.asan`) or add a subdir (e.g., `__extra_build/asan/target`).